File: prototype/Assets/microcosmicWar/Scripts/zzGameObjectCreator.js

static protected var singletonInstance:zzGameObjectCreator=null;

static function getSingleton()
{
	return singletonInstance;
}

class zzCreatorInfo
{
	//±ÿ–Î”–zzGameObjectInit
	var name:String;
	var prefab:GameObject;
}

var creatorList:zzCreatorInfo[];

//[name] = prefab
protected var creatorMap=Hashtable();


/*
["creatorName"]
["position"]
["rotation"]
*/
function create( p:Hashtable)
{
	var position=Vector3();
	var rotation=Quaternion();
	if(p.ContainsKey("position"))
		position = p["position"];
	if(p.ContainsKey("rotation"))
		rotation = p["rotation"];
	var clone = zzCreatorUtility.Instantiate( creatorMap[p["creatorName"]] , position , rotation , 0) ;
	var initObject:zzGameObjectInit = clone.GetComponent(zzGameObjectInit);
	initObject.init(p);
}

function Awake()
{
	if(singletonInstance)
		Debug.LogError("have singletonInstance");
	singletonInstance = this;
	
	for(var v:zzCreatorInfo in creatorList)
	{
		creatorMap[v.name]=v.prefab;
	}
}

File: prototype/Assets/microcosmicWar/Scripts/zz/zzGameObjectCreator.cs

using UnityEngine;
using System.Collections;

public class zzGameObjectCreator : MonoBehaviour
{


    static protected zzGameObjectCreator singletonInstance = null;

    public static zzGameObjectCreator getSingleton()
    {
        return singletonInstance;
    }

    [System.Serializable]
    public class zzCreatorInfo
    {
        //必须有zzGameObjectInit
        public string name;
        public zzGameObjectInit prefab;
    }

    public zzCreatorInfo[] creatorList;

    //[name] = prefab
    protected Hashtable creatorMap = new Hashtable();


    /*
    ["creatorName"]
    ["position"]
    ["rotation"]
    */
    public void create(Hashtable p)
    {
        Vector3 position = new Vector3();
        Quaternion rotation = new Quaternion();
        if (p.ContainsKey("position"))
            position = (Vector3)p["position"] ;
        if (p.ContainsKey("rotation"))
            rotation = (Quaternion)p["rotation"];
        GameObject clone = zzCreatorUtility.Instantiate((GameObject)creatorMap[p["creatorName"]], position, rotation, 0);
        Debug.Log(p["creatorName"] + " " + clone.name);
        zzGameObjectInit initObject = clone.GetComponent<zzGameObjectInit>();
        initObject.init(p);
    }

    void Awake()
    {
        if (singletonInstance)
            Debug.LogError("have singletonInstance");
        singletonInstance = this;

        foreach (zzCreatorInfo v in creatorList)
        {
            creatorMap[v.name] = v.prefab.gameObject;
        }
    }

}